regexpext/regexpext.go

```diff
@@ -0,0 +1,119 @@
+* limitations under the License.
+*
+*******************************************************************************/
+
+// Package regexpext provides Regexp wrapper types that automatically compile
+// regexps while they are being unmarshaled from YAML or JSON files. If the
+// compilation fails, the error will be reported as part of the Unmarshal
+// operation.
+package regexpext
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// Regexp is a regexp.Regexp that implements the Marshaler and Unmarshaler
+// interfaces for encoding/json and gopkg.in/yaml.v2, respectively. This type
+// also works with gopkg.in/yaml.v3: Even though the Unmarshaler interface has
+// changed in yaml.v3, v3 also still supports the v2 interface.
+//
+// During unmarshaling, empty string values will cause the Regexp field to be
+// set to nil. This means that empty values behave the same as absent values.
+type Regexp struct {
+	*regexp.Regexp
+}
+
+func (r Regexp) MarshalJSON() ([]byte, error)           { return marshalJSON(r.Regexp, false) }
+func (r Regexp) MarshalYAML() (any, error)              { return marshalYAML(r.Regexp, false) }
+func (r *Regexp) UnmarshalJSON(buf []byte) error        { return parseJSON(buf, &r.Regexp, false) }
+func (r *Regexp) UnmarshalYAML(u func(any) error) error { return parseYAML(u, &r.Regexp, false) }
+
+// BoundedRegexp is like Regexp, but ^ and $ anchors will automatically be
+// added to the start and end of the regexp, respectively. For example, when
+// unmarshaling the value "foo|bar" into a BoundedRegexp, the unmarshaled
+// regexp will be "^(?:foo|bar)$".
+type BoundedRegexp struct {
+	*regexp.Regexp
+}
+
+func (r BoundedRegexp) MarshalJSON() ([]byte, error)           { return marshalJSON(r.Regexp, true) }
+func (r BoundedRegexp) MarshalYAML() (any, error)              { return marshalYAML(r.Regexp, true) }
+func (r *BoundedRegexp) UnmarshalJSON(buf []byte) error        { return parseJSON(buf, &r.Regexp, true) }
+func (r *BoundedRegexp) UnmarshalYAML(u func(any) error) error { return parseYAML(u, &r.Regexp, true) }
+
+func marshalYAML(rx *regexp.Regexp, isBounded bool) (string, error) {
+	if rx == nil {
+		return "", nil
+	}
+	out := rx.String()
+	if isBounded {
+		//For round-trip safety, we need to remove the ^(?:...)$ that we added in compile().
+		//This can fail if the regex was provided in code instead of via unmarshalling.
+		if !strings.HasPrefix(out, "^(?:") || !strings.HasSuffix(out, ")$") {
+			return "", fmt.Errorf("cannot marshal BoundedRegexp because regexp is not bounded: %q", out)
+		}
+		out = strings.TrimPrefix(out, "^(?:")
+		out = strings.TrimSuffix(out, ")$")
+	}
+	return out, nil
+}
+
+func marshalJSON(rx *regexp.Regexp, isBounded bool) ([]byte, error) {
+	val, err := marshalYAML(rx, isBounded)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(val)
+}
+
+func parseJSON(buf []byte, out **regexp.Regexp, isBounded bool) error {
```

Comment:
I hope I don't forgot to ask you what the double pointers do here.

Reply:
Generally, when you have a variable of type `T` and you want another function to be able to write into it, you need to provide its address, i.e. give a `*T` argument. In this case, we want to write into a struct field that has type `*regexp.Regexp` (a non-pointer type does not work here since we want to have the option to set the field to `nil`), hence we need to provide access to it via a `**regexp.Regexp` argument.